### ISSUE TITLE
Automatically copy to clipboard

### ISSUE THREAD

**Issue by VanoDevium** [raineorshine/npm-check-updates#1172]
- [x] I have searched for [similar issues](https://github.com/raineorshine/npm-check-updates/issues)
- [x] I am using the latest version of `npm-check-updates`
- [x] I am using `node >= 14`

---

```
❯ ncu -g
[===================-] 20/21 95%
 npm-check-updates  16.0.1  →  16.0.3

ncu itself cannot upgrade global packages. Run the following to upgrade all global packages: 

npm -g install npm-check-updates@16.0.3
~ 
```

Just idea: what about automatically copy **following** to clipboard.

1. I run `ncu`
2. I past into terminal already copied string and push Enter

**Comment by raineorshine** [comment]
Hi, thanks for the suggestion. 

I wouldn't want to clobber the user's clipboard by default. Different platforms [have different means of copying to the clipboard](https://stackoverflow.com/questions/5130968/how-can-i-copy-the-output-of-a-command-directly-into-my-clipboard) from the CLI, if it supports it at all, so that is another layer of complexity. It is unconventional for a CLI program to automatically write to the clipboard, so it is probably not recommended.

Luckily, your shell is very good at processing and piping output to different places. You can store the output in a variable, then echo it and pipe it to the clipboard. Here's an example for OSX:

```sh
OUTPUT=$(ncu -g); echo $OUTPUT; echo $OUTPUT | tail -2 | pbcopy
```

A cleaner way if your shell supports it is to use `tee` to duplicate the output:

```sh
ncu -g | tee >(tail -2 | pbcopy)
```

However `tee` will strip colors from the output. To preserve colors do this:

```sh
ncu -g --color | tee >(tail -2 | sed 's/\[..m//g' | pbcopy)
```

As mentioned, replace `pbcopy` with the appropriate copy-to-clipboard utility supported by your platform/shell.

**Comment by FWDekker** [comment]
I like this idea, but agree that the clipboard interface isn't great to work with.

My concern with your input-parsing solution is that it depends on the output format of `ncu`; if even a single line is added at the end then the parsing no longer works. Unless you can confirm that the exact output format is part of your API and you increase the major version any time the output format changes (which I don't think you should do), I'm wary of using the input-parsing solution.

As an alternative, what do you think of either
1. adding an `--output-updateables-only` option; if enabled, the only thing written to stdout by `ncu` is a list of the packages that can be updated, or
2. adding a `--write-updateables-to [filename]` option; if enabled, `ncu` writes the list of packages that can be updated to `[filename]`?

Basically, these options just formalise the guarantee that the output is in a specific format and is part of the API.

**Comment by raineorshine** [comment]
Yes, the output parsing is definitely fragile. 

The normal output is intended to primarily be human-readable, but with the command spelled out to make it a bit easier to copy and execute. 

For machine-readable output, specify `--jsonUpgraded`. This will output the upgraded packages in JSON. From there you can do more complex processing. This is preferred over increasing the surface area of the API by adding options to control the output.

**Comment by vanodevium** [comment]
JSON format isn't common for CLI usage too, imho

**Comment by raineorshine** [comment]
That's a fair point. In that case, I would probably add `--format lines` or `--format shell` for a machine-readable, line-base output that is considered part of the API.

**Comment by vanodevium** [comment]
Nice! 

`--format {lines|shell|comma}`

**Comment by FWDekker** [comment]
> Yes, the output parsing is definitely fragile.
> 
> The normal output is intended to primarily be human-readable, but with the command spelled out to make it a bit easier to copy and execute.
> 
> For machine-readable output, specify `--jsonUpgraded`. This will output the upgraded packages in JSON. From there you can parse it with a tool like [jq](https://stedolan.github.io/jq/) and do more complex processing. This is preferred over increasing the surface area of the API by adding options to control the plain text output.

Awesome! Using [jq](https://stedolan.github.io/jq/) I could easily parse the JSON on command line. To check and immediately upgrade, I can just do
```shell
sudo npm -g i $(sudo ncu -g --jsonUpgraded | jq -r 'keys | join(" ")')
```
I'll probably write something so that it first prompts for confirmation before it starts installing it, but this snippet is the gist of it :-)

**Comment by vanodevium** [comment]
jq is a great tool but needs to be installed separately.
We are considering a format that can be parsed on any unix system.

Moreover, --jsonUpgraded is not going anywhere ;)

**Comment by FWDekker** [comment]
> jq is a great tool but needs to be installed separately.
> We are considering a format that can be parsed on any unix system.

Fair enough. Since `ncu` obviously requires `npm`, and `npm` requires NodeJS, how about using Node's `-p` switch to execute JS to parse the JSON? So you could do the following:

```shell
sudo npm -g i $(node -p "Object.keys(JSON.parse(process.argv[1])).join(' ')" -- "$(sudo ncu -g --jsonUpgraded)")
```

> Moreover, --jsonUpgraded is not going anywhere ;)

What do you mean by this?

**Comment by vanodevium** [comment]
I mean

`sudo npm i -g $(ncu -g --format shell)`

is more readable

**Comment by FWDekker** [comment]
Absolutely agree. I think it'd be nice. But until then future readers will have a workaround ^^

**Comment by vanodevium** [comment]
> That's a fair point. In that case, I would probably add `--format lines` or `--format shell` for a machine-readable, line-base output that is considered part of the API.

@raineorshine 
Are there any news about format options?
